---
title: ' RNAseq Analysis'
author: "Hernan Lorenzi"
date: "04/20/2022"
output:
  html_document:
    df_print: paged
---

# Load libraries
```{r load libraries}
suppressMessages(library("pheatmap"))
suppressMessages(library("EnhancedVolcano"))
suppressMessages(library("ggplot2"))
suppressMessages(library("ggpubr"))
suppressMessages(library("DESeq2"))
suppressMessages(library("stringr"))
suppressMessages(library("biomaRt"))
suppressMessages(library("tidyverse"))
suppressMessages(library("pcaExplorer"))
suppressMessages(library("VennDiagram"))
suppressMessages(library("clusterProfiler"))
suppressMessages(library("GOSemSim"))
suppressMessages(library("ggsci"))
suppressMessages(library("viridis"))
suppressMessages(library("ggrepel"))
suppressMessages(library("RColorBrewer"))
suppressMessages(library("msigdbr"))
suppressMessages(library("cowplot"))
suppressMessages(library("enrichplot"))
#suppressMessages(library("ReactomePA"))
suppressMessages(library("ggupset"))
suppressMessages(library("broom"))
suppressMessages(library("ggraph"))
```

# Define functions
```{r aux functions}

# Load auxyliary functions
source(file = "./01_aux_rnaseq_functions.R")

# Load enrichment functions
source(file = "./02_Gene_enrichment_functions.R")

```


# Load read counts data
```{r Loading data}
all <- read.delim2("./data/read_counts", sep = "\t", header = TRUE, row.names = 1, comment.char = c("#") )

# remove unused columns
all <- all[,6:ncol(all)]

# Make sure read counts are numeric and rounded to 0 decimals
all.tmp <- as.data.frame(lapply(all, function(x){ round(as.numeric(x), digits = 0)} ))
rownames(all.tmp) <- rownames(all)
all <- all.tmp

# adjust colnames
colnames(all) <- str_remove(string = str_remove(string = colnames(all), pattern = ".sorted.dedup.bam"), pattern = "results.04dedup.")

#Remove all zero rows
all <- remove_all_zero_rows(all, min_total_count = 0)

```

# Make metadata table
```{r}
metadata <- as.data.frame(str_split(string = str_remove(string = colnames(all), pattern = "LF82_"), pattern = "_", simplify = T))
metadata['sample_id'] <- colnames(all)
rownames(metadata) <- metadata$sample_id

# Add total read counts and sample id columns to metadata
metadata$read_counts <- colSums(all)
colnames(metadata) <- c('Group','Replicate','Sample_name','Read_count')
```


# Normalize data to TPMs to run some comparative analysis across samples
```{r}
# Using annotation version GRCm39 (current)
#all.tpm <- normalize_by_TPM(counts.df = all)
```

###########################################
# Analysis of expression data using DESeq2
###########################################

```{r deseq2}
# Convert metadata to factors
for (variable in c("Group", "Replicate","Sample_name")){
  metadata[,variable] <- as.factor(metadata[,variable])  
}

```

# Analysis of Dataset

```{r Dataset one}
# Generate DESeq2 object for NS and ST condition ONLY. We could potentially add Read_counts as either a covariate or as a factor (by grouping Read counts into Low and High count bins). Ideally, this should be done adding Read_counts as a random effect using mixed models, that can only be applied in Limma package. Adding it as a fixed effect in the design formula might lower the power of the analysis.

dir.create(path = "./Plots", showWarnings = FALSE)

# Create DESeq object
dds.all <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~ Group)

# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "Sample_name", 
          y = "Read_count",
          x.text.angle = 90,
          fill = "Group", 
          title = "Total read counts", 
          ylab = "Read count",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc") 
ggsave2("Plots/barplot_read_counts.pdf", plot = p)
print(p)
```


```{r}
# Normalize counts
vsd.one <- vst(dds.all, blind=FALSE)
rlog.one <- rlog(dds.all, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(counts(dds.all)) >= 20
dds.all <- dds.all[keep,]

# Calculate distances between samples
sampleDists <- dist(t(assay(vsd.one)))

# Plot inter-sample distances
old.par <- par(no.readonly=T)

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rlog.one$Sample_name, rlog.one$Genotype, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
p.pheatmap <- pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

ggsave2(filename = "unsupervised_clustering_rnaseq_profile_20plus_reads.pdf", plot = p.pheatmap, path = "./Plots")
print(p.pheatmap)
```
Heatmap: Samples do not separates by genotype. Maybe there is a gender issue? Do we have gender information?


```{r}
dds_res <- list()

dds_res <- dds.all #[ , dds.all$Tissue=="all_data"]

rlog_res <- list()
rlog_res <- rlog(dds_res, blind=FALSE)

# PCA
rlog.one <- rlog_res

# PC1 - PC2
PC_coords = c(1,2)
pcaData <- plotPCA_anycoord(object = rlog.one, intgroup=c("Group"), returnData=TRUE, PC_coords = PC_coords)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords

p12 <- ggplot(data = pcaData, aes_string(x = "PC1", y = "PC2", color = "Group", label = "name")) +
  geom_point(size = 3) + 
  xlab(paste0("PC",PC_coords[1],": ", round(percentVar[1]), "% variance")) + 
  ylab(paste0("PC",PC_coords[2],": ", round(percentVar[2]), "% variance")) + 
  coord_fixed() + 
  geom_text_repel(size=3) +
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + # Remove grid
  scale_color_brewer(type = "div", palette = "Set1") + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) +
  labs(title = my_tissue) 
print(p12)
ggsave(paste0("Plots/pca_PC",PC_coords[1],PC_coords[2],"_Group.pdf"), plot = p12)

# PC2 - PC3
PC_coords = c(2,3)
pcaData <- plotPCA_anycoord(object = rlog.one, intgroup=c("Group"), returnData=TRUE, PC_coords = PC_coords)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords
p1 <- ggplot(data = pcaData, aes_string(x = "PC1", y = "PC2", color = "Group", label = "name")) +
  geom_point(size = 3) + 
  xlab(paste0("PC",PC_coords[1],": ", round(percentVar[1]), "% variance")) + 
  ylab(paste0("PC",PC_coords[2],": ", round(percentVar[2]), "% variance")) + 
  coord_fixed() + 
  geom_text_repel(size=3) +
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + # Remove grid
  scale_color_brewer(type = "div", palette = "Set1") + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) +
  labs(title = my_tissue) 
print(p1)
ggsave(paste0("Plots/pca_PC",PC_coords[1],PC_coords[2],"_Group.pdf"), plot = p1)

# PC1 - PC3
PC_coords = c(1,3)
pcaData <- plotPCA_anycoord(object = rlog.one, intgroup=c("Group"), returnData=TRUE, PC_coords = PC_coords)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords

p13 <- ggplot(data = pcaData, aes_string(x = "PC1", y = "PC2", color = "Group", label = "name")) +
  geom_point(size = 3) + 
  xlab(paste0("PC",PC_coords[1],": ", round(percentVar[1]), "% variance")) + 
  ylab(paste0("PC",PC_coords[2],": ", round(percentVar[2]), "% variance")) + 
  coord_fixed() + 
  geom_text_repel(size=3) +
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + # Remove grid
  scale_color_brewer(type = "div", palette = "Set1") + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) +
  labs(title = my_tissue) 
print(p13)
ggsave(paste0("Plots/pca_PC",PC_coords[1],PC_coords[2],"_Group.pdf"), plot = p13)

```
PCA analysis shows that samples tend to separate by genotype, but the clustering is not very strong.

### resultsNames(dds)

# Filtering out poorly-expressed genes (less than 20 reads across all samples) 
```{r filtering genes based on total counts of reads}

# Keep genes with at least 10 reads total across samples
keep <- rowSums(counts(dds_res)) >= 20
dds_res <- dds_res[keep,]

```


# Using groups instead of interactions
```{r}
ensembl_to_symbol <- read_delim(file = "./data/gene_names.txt", delim = "\t", col_names = c("Ensembl_ID","gene_name"))

# Save sorted files as a list
DE_results <- list()
geneids.DE <- list()

# Define function for processing and saving result tables
sort_and_write_res_table <- function(result_table, file_name){
  dir.create(path = "./DE", showWarnings = FALSE)
  # Sort genes by (padj)
  result_table_sorted <- result_table[order(result_table$padj, decreasing = FALSE),]
  # Add gene symbols
  gene_list <- rownames(result_table_sorted)
  symbol_list <- ensembl_to_symbol$gene_name[match(gene_list, ensembl_to_symbol$Ensembl_ID)]
  df <-as.data.frame(cbind(result_table_sorted, Gene_name = symbol_list))
  
  # Write sorted table to file
  write.table(df, file = paste0("./DE/",file_name,".txt"), 
            sep = "\t", col.names=NA)
  return(df)
}
```

```{r}
# Calculate DE for all_data samples
#design(dds.rnaseA) <- ~Treatment # Removid Read.depth from formula given that all samples are Read.depth = High

dds_res$Group <- relevel(dds_res$Group, "WT")
dds_res <- DESeq(dds_res)
my_contrasts <- resultsNames(dds_res)


# Using lfcShrink instead of results to reduce high Log2FC bias of genes with low expression
res_genotype_SD105_vs_WT <- lfcShrink(dds_res, coef = my_contrasts[2], type = "ashr", )
res_genotype_SD106_vs_WT <- lfcShrink(dds_res, coef = my_contrasts[3], type = "ashr", )
res_genotype_SD107_vs_WT <- lfcShrink(dds_res, coef = my_contrasts[4], type = "ashr", )

summary(res_genotype_SD105_vs_WT, alpha = 0.05)
summary(res_genotype_SD106_vs_WT, alpha = 0.05)
summary(res_genotype_SD107_vs_WT, alpha = 0.05)

# Sort results by Log2FC
res_genotype_SD105_vs_WT_sorted <- sort_and_write_res_table(result_table = res_genotype_SD105_vs_WT, file_name = paste0("DE_",my_tissue,"_SD105_vs_WT")) 
res_genotype_SD106_vs_WT_sorted <- sort_and_write_res_table(result_table = res_genotype_SD106_vs_WT, file_name = paste0("DE_",my_tissue,"_SD106_vs_WT")) 
res_genotype_SD107_vs_WT_sorted <- sort_and_write_res_table(result_table = res_genotype_SD107_vs_WT, file_name = paste0("DE_",my_tissue,"_SD107_vs_WT")) 

# Save sorted files as a list
DE_results[[my_tissue]]  <- res_genotype_KO_vs_WT_sorted


# Save list of genes significantly DE as a list
geneids.DE[[my_tissue]] <- rownames(subset(res_genotype_KO_vs_WT_sorted, padj <= 1))
  

```


```{r}
print(sessionInfo())
```